libpromises/processes_select_proc.c

```diff
@@ -0,0 +1,809 @@
+ * Here we examine likely places under "/dev".
+ * In this initial implementation we only look for "/dev/pts/nnn"/.
+ *
+ * The "/proc" data has given us the major/minor device "ttyn".
+ * We look through our possible candidate list in "/dev" for a match.
+ */
+static bool LoadProcTTY(JsonElement *pdata, int ttyn)
+{
+    char ttyname[CF_MAXVARSIZE];
+    struct stat statbuf;
+    int ttymajor = major(ttyn);
+    int ttyminor = minor(ttyn);
+
+    JsonObjectAppendInteger(pdata, JPROC_KEY_TTY, ttyn);
+
+    /*
+     * From kernel "devices.txt", character-major 136-143 inclusive are "pty".
+     * We do a quick sanity-check on "/dev/pts/nnnn".
+     * If it looks OK, we set the name "pts/nnn".
+     */
+    if (ttymajor >= 136 && ttymajor <= 143)
+    {
+        snprintf(ttyname, CF_MAXVARSIZE, "/dev/pts/%d", ttyminor);
+        if (stat(ttyname, &statbuf) >= 0)
+        {
+            JsonObjectAppendString(pdata, JPROC_KEY_TTYNAME, &ttyname[5]);
+            return true;
+        }
+    }
+
+    /*
+     * Could probably check for other things; "/dev/tty<n>" springs to mind.
+     * But for now, fail safely.
+     */
+
+    return false;
+}
+
+static JsonElement *LoadProcStat(pid_t pid)
+{
+    char statfile[CF_MAXVARSIZE];
+    int fd, len;
+    JsonElement *pdata;
+
+    pdata = JsonObjectCreate(12);
+
+    /* get uid from file 'status' */
+    if (! LoadProcUid(pdata, pid))
+    {
+        return pdata;
+    }
+
+    /* extract command line info and store in 'pdata' */
+    LoadProcCmd(pid, pdata);
+
+    /* open the 'stat' file */
+    snprintf(statfile, CF_MAXVARSIZE, "/%s/%d/stat", PROCDIR, pid);
+    fd = open(statfile, O_RDONLY, 0);
+    if (fd < 0) {
+        return pdata;
+    }
+
+    /* read the 'stat' file into a buffer */
+    char statbuf[CF_MAXVARSIZE];
+    len = read(fd, statbuf, CF_MAXVARSIZE -1);
+    close(fd);
+    if (len <=0) {
+        return pdata;
+    }
+
+    /*
+     * Big picture: "stat" is:
+     *    pid (some text here) S n1 n2 n3 ...
+     * We already know the pid. And the text is a variant of command (known).
+     * We are interested in the state char 'S' and numbers n1, n2, n3, ...
+     */
+
+    /* Find closing ')' and place pointer on the char 'S'. */
+    char *pt;
+    pt = strrchr(statbuf, ')');
+    pt++;
+    pt++;
+
+    /* Extract the 'stat' values.  May vary with OS/release. */
+
+#if defined(__linux__)
+
+    char pstate[2];
+    pid_t ppid;
+    pid_t pgid;
+    int ttyn;
+    long utime;
+    long stime;
+    long cpusecs;
+    long priority;
+    long nice;
+    long threads;
+    long long starttime;
+    ulong vsize;
+    long rss;
+
+    pstate[1] = '\0';
+    sscanf(pt,
+      "%c "
+      "%d %d %*d %d %*d %*u "
+      "%*u %*u %*u %*u "
+      "%lu %lu %*d %*d "
+      "%ld %ld %ld %*d "
+      "%llu %lu %ld %*d ",
+      &pstate[0],
+      &ppid, &pgid, &ttyn,
+      &utime, &stime,
+      &priority, &nice, &threads,
+      &starttime, &vsize, &rss);
```

Comment:
Taken from travis valgrind output:

```
==2109== Use of uninitialised value of size 8
==2109==    at 0x66D4550: rawmemchr (rawmemchr.S:25)
==2109==    by 0x66BAFF1: _IO_str_init_static_internal (strops.c:41)
==2109==    by 0x66AE266: vsscanf (iovsscanf.c:40)
==2109==    by 0x66A8916: sscanf (sscanf.c:32)
==2109==    by 0x4ED388C: LoadProcStat (processes_select_proc.c:679)
==2109==    by 0x4ED4335: LoadProcessTable (processes_select_proc.c:788)
==2109==    by 0x40D6AA: KeepAgentPromise (cf-agent.c:1667)
==2109==    by 0x4EA9293: ExpandPromiseAndDo (expand.c:215)
==2109==    by 0x4EA9293: ExpandPromise (expand.c:283)
==2109==    by 0x40DF6A: ScheduleAgentOperations (cf-agent.c:1414)
==2109==    by 0x41AD8D: VerifyMethod (verify_methods.c:172)
==2109==    by 0x41B13C: VerifyMethodsPromise (verify_methods.c:74)
==2109==    by 0x40D965: KeepAgentPromise (cf-agent.c:1729)
==2109==  Uninitialised value was created by a stack allocation
==2109==    at 0x4ED3560: LoadProcStat (processes_select_proc.c:616)
```

At this point, it's a little confusing to see what's happening, but I think this is because `pt` points to an unitialized part of statbuf (which is stack allocated in this funcion, LoadProcStat()), so should be fixed when you address the comments above.